### ISSUE TITLE
mlx_lm.server: seed parameter silently ignored

### ISSUE THREAD

**Issue by guygrigsby** [ml-explore/mlx-lm#1245]
## `mlx_lm.server`: `seed` parameter silently ignored

Chat-completions endpoint accepts `seed` and routes it through `mx.random.seed`, but output is the same across different seed values for the same prompt. mlx-lm 0.31.3, M-series, reproduced on Qwen2.5 and Llama-3.1 4 bit quants.

## Repro

Start server:

```bash
MODEL_PATH=/path/to/your/mlx/model
mlx_lm.server --model "$MODEL_PATH" --port 1234
```

Send three requests with different seeds (same shell, `$MODEL_PATH` carries over):

```bash
for s in 11111 22222 33333; do
  curl -sS http://127.0.0.1:1234/v1/chat/completions \
    -H 'Content-Type: application/json' \
    -d "{\"model\":\"$MODEL_PATH\",\"temperature\":1,\"max_tokens\":25,\"seed\":$s,\"stream\":false,\"messages\":[{\"role\":\"user\",\"content\":\"Tell me a one-sentence story about a cat.\"}]}" \
  | jq -r '.choices[0].message.content'
done
```
```
bash repro.sh
The tabby cat's reflection in the dark window was the only companion she hadn't driven away that evening.
The tabby cat's reflection in the dark window was the only companion she hadn't driven away that evening.
The tabby cat's reflection in the dark window was the only companion she hadn't driven away that evening.
```

All three responses come back byte-identical.

## Expected

Different seeds should follow different sampling trajectories which should provide different outputs (at temperature > 0).

## My thoughts why

`server.py:1191` reads `seed` from the body into `self.seed`. At `server.py:956-957` it's applied:

```python
if args.seed is not None:
    mx.random.seed(args.seed)
```

That call executes. Sampling lives at `sample_utils.py:267-282`:

```python
@partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)
def categorical_sampling(logits, temp):
    return mx.random.categorical(logits * (1 / temp))
```

~Hypothesis: `mx.compile` with `inputs=mx.random.state` captures the state object at decoration time, so subsequent `mx.random.seed(...)` calls don't propagate into what the compiled function reads. Effectively, sampling becomes deterministic against whatever state existed at module import.~
### Update

Initial suspicion was that the `mx.compile` decorator captured a stale `mx.random.state` reference. Not actually it: running `mx.random.seed(N) -> mx.random.categorical(...)` directly with three different seeds produces three different draws, both with and without the decorator. So it looks like the bug is somewhere downstream of the seed call but upstream of (or alongside) the actual sampling.
## Also tried

- Removed `seed` from the body, ran the same prompt twice. Identical output. State isn't naturally advancing across requests either.
- Same prompt at temperature 0.5 / 1.5 / 3.0. Outputs differ. Sampling runs, temperature is honored. Seed/state path is broken.
- Different prompts at fixed seed. Outputs differ. Not response-caching.

## Workaround

Vary something in the prompt per request (a hidden tag in the system message works). At least breaks the unintended determinism.

**Comment by IsmaelMartinez** [comment]
Reproduced this here on mlx_lm 0.31.3 (Apple Silicon): seeds 1/2/3 at temperature 1.0 return byte-identical output, and the same model varies fine on another backend, so it is the server, not the weights.

The fix in #1331 sorts it. Validated with a before/after seed repro plus its added tests (30 passed). Details over there.

**Comment by sashko-zakharchuk** [comment]
I think this is `mx.compile` capturing thread-local random state at import time.

`apply_xtc` and `categorical_sampling` in `sample_utils.py` are decorated with `@partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)`. Decorator arguments get evaluated once, at import, on whichever thread does the import, and `mx.random.state` is thread local. So both functions end up pinned to the importing thread's state. Call them from another thread and sampling just stops moving: same inputs give the same value forever, and `mx.random.seed` on that thread does nothing to it.

`mlx_lm.server` does all its generation on `ResponseGenerator._generate`, which is a worker thread (`server.py:430`). That is why identical requests come back identical at any temperature and `seed` appears to be ignored. `mlx_lm.generate` and `mlx_lm.chat` run on the main thread, so they are fine, which I assume is why nobody hit this sooner.

Smallest repro I could get, no model or server needed, on mlx 0.32.0:

```python
import threading

import mlx.core as mx

from mlx_lm.sample_utils import make_sampler

logits = mx.random.normal((1, 4096))
mx.eval(logits)
sampler = make_sampler(temp=1.0)


def draw(where):
    tokens = [int(sampler(logits).item()) for _ in range(32)]
    print(f"{where}: {len(set(tokens))} distinct out of 32")


draw("main thread  ")
t = threading.Thread(target=draw, args=("worker thread",))
t.start()
t.join()
```

```
main thread  : 32 distinct out of 32
worker thread: 1 distinct out of 32
```

Uncompiled `mx.random.categorical` on that same worker thread varies fine, so it is not the RNG itself.

One thing worth flagging before anyone writes a patch: this looks already fixed upstream by ml-explore/mlx#3828, merged 2026-07-20, which resolves the captured state to the calling thread's key and adds `test_compile_rng_across_threads`. It has not shipped in a release yet (v0.32.0 predates it by about two weeks). I built MLX main and the snippet gives 32 distinct on both threads. So bumping `MIN_MLX_VERSION` once that lands might be all this needs.
